### PR TITLE
Add fees syscall to expose cluster fees into programs

### PR DESCRIPTION
**Pull request by mvines**
TODO:
* [x] bank.rs-level unit test

### CODE REVIEW

**Comment by mvines** [comment]
@rob-solana - I'm just cargo culting your work in slot_hash.rs, does this look like about what you'd expect?

File: sdk/src/syscall/fees.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct Fees {
+    pub fee_calculator: FeeCalculator,
+}
+
+impl Fees {
+    pub fn from(account: &Account) -> Option<Self> {
+        account.state().ok()
+    }
+    pub fn to(&self, account: &mut Account) -> Option<()> {
```
Comment (rob-solana) [inline sdk/src/syscall/fees.rs]: I don't really like the names I made up here...  I was thinking about a `with(FnMut)` thing at some point to cover these really short serialize/deserialize lifetimes

**Comment by rob-solana** [comment]
yep

**Comment by rob-solana** [comment]
at some point syscall accounts have to be special-cased...

we don't want them to be part of the bank delta, i.e. we don't want store(syscall) to count as a vote-able modification.  my slot_hashes stuff is exempted as a side effect of being after we hash() and freeze()

**Comment by rob-solana** [comment]
2 places I can think of where syscalls would need to be special-cased:  is_voteable() code and the hash() code

**Comment by mvines** [comment]
For dynamic fees, I think we'd adjust them at the end of an epoch based on the cluster activity from the previous N epochs.  Hashes-per-tick probably adjust in the same way, always at an epoch boundary.    

But so long as syscall accounts are always updated deterministically across all nodes, why would we need to special case them?

**Comment by rob-solana** [comment]
`store()` marks a bank as having a delta...  if we're not careful with that, we can end up with votable banks that shouldn't be votable, and the hash() stuff is just saying that we want hash() to behave like everything else (ignore synthesized accounts)

**Comment by mvines** [comment]
@rob-solana, I think this is ready for review.
* The fee schedule is fixed (until #4167) so the `store()` issues you raised shouldn't apply just yet
* Want me to wait until #4488 lands and use that model instead?

**Comment by rob-solana** [comment]
no need to wait for #4488 , I'm happy to go fix everybody using the old forms

**Pull request merged**